sh_scrapy/extension.py

```diff
@@ -46,32 +48,32 @@ def item_scraped(self, item, spider):
         self._write_item(item)
 
     def spider_closed(self, spider, reason):
-        # flush item writer
-        self.hsref.job.items.flush()
-        # update outcome
-        self.hsref.job.metadata.update(close_reason=reason)
-        self.hsref.job.metadata.save()
+        self.pipe_writer.flush()
+        self.pipe_writer.set_outcome(reason)
 
 
 class HubstorageMiddleware(object):
 
     def __init__(self):
         self._seen = WeakKeyDictionary()
         self.hsref = hsref.hsref
+        self.pipe_writer = pipe_writer
+        self.request_id_sequence = itertools.count()
 
     def process_spider_input(self, response, spider):
         parent = response.meta.get('_hsparent')
-        riq = self.hsref.job.requests.add(
-            parent=parent,
-            url=response.url,
-            status=response.status,
-            method=response.request.method,
-            rs=len(response.body),
-            duration=response.meta.get('download_latency', 0) * 1000,
-            ts=time.time() * 1000,
-            fp=request_fingerprint(response.request),
-        )
-        self._seen[response] = riq
+        request_object = {
+            'url': response.url,
+            'status': int(response.status),
```

Comment:
Out of curiosity, but why do you need conversion here? 
It was standardised in Scrapy more than [8 years ago](https://github.com/scrapy/scrapy/commit/f63a661320805e80b0a078752e7a4f6218141418#diff-1f09c5b98974d24900544c14f968069eR34).